src/providers/krb5/krb5_common.c

```diff
@@ -1141,38 +1141,26 @@ void krb5_finalize(struct tevent_context *ev,
                    void *siginfo,
                    void *private_data)
 {
+    struct be_ctx *be_ctx;
+
+    be_ctx = talloc_get_type(private_data, struct be_ctx);
+    talloc_free(be_ctx);
```

Comment:
Freeing `be_ctx` seems wrong on this level. Shouldn't you rather free some Kerberos context?

Reply:
We encountered an issue with TCP connection and related destructor.

The code is a bit complicated, lot of references between structures. I'm not 100% sure if I read the code correctly. By my understanding be_ctx holds connections (trough failover code).

@pbrezina am I right?

Reply:
Reading backtrace in [comment 1](https://bugzilla.redhat.com/show_bug.cgi?id=1672584#c1), the crash happens when it tries to call `ldap_unbind_ext` on a handle that was probably already released automatically at exit. Freeing `be_ctx` manually will make sure that everything is cleared before calling atexit callbacks, but it should be done in proper place.

For example, you can create something like:
```c
void be_orderly_shutdown(int status, struct be_ctx *be_ctx)
{
    talloc_free(be_ctx);
    orderly_shutdown(status);
}
```